pkg/ring/flush.go

```diff
@@ -0,0 +1,30 @@
+package ring
+
+import "context"
+
+// FlushTransferer controls the shutdown of an instance in the ring.
+type FlushTransferer interface {
+	StopIncomingRequests()
+	Flush()
+	TransferOut(ctx context.Context) error
+}
+
+// NoopFlushTransferer is a FlushTransferer which does nothing and can
+// be used in cases we don't need one
+type NoopFlushTransferer struct{}
+
+// NewNoopFlushTransferer makes a new NoopFlushTransferer
```

Comment:
The ruler has identical functionality in `pkg/ruler/lifecycle.go` it would be cool to delete that and reuse this instead.

Reply:
The ruler `Flush()` is not a noop. Currently the `NoopFlushTransferer` is a adopt all or nothing: how would you see reusing it in the ruler?

Reply:
True, I forgot the change to make the ruler flush has not been merged yet:
https://github.com/cortexproject/cortex/pull/1571/files#diff-95b5a43683ab83429e93fef5c5daf87fL26

Reply:
No problem. We can use the `NoopFlushTransferer` in the ruler as soon as the PR 1571 will be merged.